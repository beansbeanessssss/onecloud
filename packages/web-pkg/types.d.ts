import { Ability } from 'web-client/src/helpers/resource/types'
import { ClientService, LoadingService, PreviewService, ArchiverService } from './src/services'

export * from './src'

declare module 'vue' {
  interface ComponentCustomProperties {
    $ability: Ability
    $archiverService: ArchiverService
    $clientService: ClientService
    $loadingService: LoadingService
    $previewService: PreviewService
  }
}
