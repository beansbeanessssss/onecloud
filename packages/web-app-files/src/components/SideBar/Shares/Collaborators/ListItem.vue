<template>
  <div
    :data-testid="`collaborator-${isUser || isSpace ? 'user' : 'group'}-item-${
      share.collaborator.name
    }`"
    class="files-collaborators-collaborator oc-flex oc-flex-middle oc-py-xs oc-flex-between"
  >
    <div class="oc-width-1-1 oc-flex oc-flex-middle" style="gap: 10px">
      <avatar-image
        v-if="isUser || isSpace"
        :userid="share.collaborator.name"
        :user-name="share.collaborator.displayName"
        :width="48"
        class="files-collaborators-collaborator-indicator"
      />
      <oc-avatar-item
        v-else
        :width="48"
        icon-size="medium"
        :icon="shareTypeIcon"
        :name="shareTypeKey"
        class="files-collaborators-collaborator-indicator"
      />
      <div class="oc-text-truncate oc-width-1-1">
        <div class="oc-flex oc-flex-middle">
          <span v-oc-tooltip="shareDisplayNameTooltip" class="oc-text-truncate oc-m-rm">
            <span
              aria-hidden="true"
              class="files-collaborators-collaborator-name"
              v-text="shareDisplayName"
            />
          </span>
          <span class="oc-invisible-sr" v-text="screenreaderShareDisplayName" />
          <oc-drop
            ref="accessDetails"
            class="share-access-details-drop"
            mode="manual"
            :target="`#edit-drop-down-${editDropDownToggleId}`"
          >
            <h5 v-translate class="oc-text-bold oc-m-rm">Access details</h5>
            <dl class="oc-mt-s">
              <dt v-if="shareAdditionalInfo" v-translate class="oc-text-muted oc-mb-s">Addition</dt>
              <dd
                v-if="shareAdditionalInfo"
                class="files-collaborators-collaborator-additional-info"
                v-text="shareAdditionalInfo"
              />
              <dt v-translate class="oc-text-muted">Type</dt>
              <dd class="files-collaborators-collaborator-share-type" v-text="shareTypeText" />
            </dl>
          </oc-drop>
        </div>
        <div class="oc-m-rm oc-flex oc-flex-middle oc-flex-between">
          <div>
            <div v-if="canEditOrDelete" class="oc-flex oc-flex-nowrap oc-flex-right oc-flex-middle">
              <role-dropdown
                :resource="highlightedFile"
                :dom-selector="shareDomSelector"
                :existing-permissions="share.customPermissions"
                :existing-role="share.role"
                :allow-share-permission="hasResharing || isSpace"
                class="files-collaborators-collaborator-role"
                @optionChange="shareRoleChanged"
              />
            </div>
            <div v-else-if="share.role">
              <span class="oc-mr-xs" v-text="share.role.label" />
            </div>
          </div>
          <div class="oc-flex oc-flex-between oc-flex-middle oc-pl-s">
            <span v-if="hasExpirationDate">
              <oc-icon
                v-oc-tooltip="expirationDate"
                class="files-collaborators-collaborator-expiration"
                data-testid="recipient-info-expiration-date"
                :aria-label="expirationDate"
                name="calendar"
                fill-type="line"
              />
              <span class="oc-invisible-sr" v-text="screenreaderShareExpiration" />
            </span>
            <div v-if="sharedParentRoute" class="oc-resource-indicators oc-text-truncate">
              <router-link
                v-oc-tooltip="$gettext('Navigate to parent folder')"
                class="parent-folder oc-text-truncate"
                :to="sharedParentRoute"
              >
                <span class="text" v-text="$gettext('via')" />
                <oc-icon name="folder-2" size="small" fill-type="line" class="oc-px-xs" />
                <span class="text oc-text-truncate" v-text="sharedParentDir" />
              </router-link>
            </div>
            <edit-dropdown
              :id="`edit-drop-down-${editDropDownToggleId}`"
              class="files-collaborators-collaborator-edit"
              data-testid="collaborator-edit"
              :expiration-date="share.expires ? share.expires : null"
              :share-category="shareCategory"
              :can-edit-or-delete="canEditOrDelete"
              @expirationDateChanged="shareExpirationChanged"
              @removeShare="removeShare"
              @showAccessDetails="showAccessDetails"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { mapGetters, mapActions, mapState } from 'vuex'
import Mixins from '../../../../mixins'
import { DateTime } from 'luxon'

import EditDropdown from './EditDropdown.vue'
import RoleDropdown from './RoleDropdown.vue'
import { SharePermissions, ShareTypes } from '../../../../helpers/share'
import { useCapabilityFilesSharingResharing } from 'web-pkg/src/composables'
import { extractDomSelector } from '../../../../helpers/resource'
import { defineComponent } from '@vue/composition-api'
import { useGraphClient } from 'web-client/src/composables'
import * as uuid from 'uuid'

export default defineComponent({
  name: 'ListItem',
  components: {
    EditDropdown,
    RoleDropdown
  },
  mixins: [Mixins],
  props: {
    share: {
      type: Object,
      required: true
    },
    modifiable: {
      type: Boolean,
      default: false
    },
    sharedParentRoute: {
      type: Object,
      default: null
    }
  },
  setup() {
    return {
      ...useGraphClient(),
      hasResharing: useCapabilityFilesSharingResharing()
    }
  },
  computed: {
    ...mapGetters('Files', ['highlightedFile']),
    ...mapState(['user']),

    shareType() {
      return ShareTypes.getByValue(this.share.shareType)
    },

    shareTypeIcon() {
      return this.shareType.icon
    },

    shareTypeKey() {
      return this.shareType.key
    },

    shareDomSelector() {
      if (!this.share.id) {
        return undefined
      }
      return extractDomSelector(this.share.id)
    },

    isGroup() {
      return this.shareType === ShareTypes.group
    },

    isUser() {
      return this.shareType === ShareTypes.user
    },

    isSpace() {
      return this.shareType === ShareTypes.space
    },

    shareTypeText() {
      return this.$gettext(this.shareType.label)
    },

    shareCategory() {
      return ShareTypes.isIndividual(this.shareType) ? 'user' : 'group'
    },

    shareDisplayName() {
      if (this.user.id === this.share.collaborator.name) {
        return this.$gettextInterpolate(this.$gettext('%{collaboratorName} (me)'), {
          collaboratorName: this.share.collaborator.displayName
        })
      }
      return this.share.collaborator.displayName
    },

    shareAdditionalInfo() {
      return this.share.collaborator.additionalInfo
    },

    shareDisplayNameTooltip() {
      return (
        this.shareDisplayName + (this.shareAdditionalInfo ? `(${this.shareAdditionalInfo})` : '')
      )
    },

    screenreaderShareDisplayName() {
      const context = {
        displayName: this.share.collaborator.displayName,
        ...(this.share.collaborator.additionalInfo && {
          additionalInfo: this.share.collaborator.additionalInfo
        })
      }
      let translated = this.$gettext('Share receiver name: %{ displayName }')
      if (this.shareAdditionalInfo) {
        translated = this.$gettext('Share receiver name: %{ displayName } (%{ additionalInfo })')
      }
      return this.$gettextInterpolate(translated, context)
    },

    shareExpirationText() {
      const translated = this.$gettext('Expires %{ expiryDateRelative }')
      return this.$gettextInterpolate(translated, {
        expiryDateRelative: this.expirationDateRelative
      })
    },

    screenreaderShareExpiration() {
      const translated = this.$gettext('Share expires %{ expiryDateRelative } (%{ expiryDate })')
      return this.$gettextInterpolate(translated, {
        expiryDateRelative: this.expirationDateRelative,
        expiryDate: this.expirationDate
      })
    },

    canEditOrDelete() {
      return this.modifiable
    },

    hasExpirationDate() {
      return this.share.expires
    },

    expirationDate() {
      return DateTime.fromJSDate(this.share.expires)
        .endOf('day')
        .setLocale(this.$language.current)
        .toLocaleString(DateTime.DATETIME_FULL)
    },

    expirationDateRelative() {
      return DateTime.fromJSDate(this.share.expires)
        .endOf('day')
        .setLocale(this.$language.current)
        .toRelative()
    },

    sharedParentDir() {
      return this.sharedParentRoute?.params?.item.split('/').pop()
    },

    shareDetailsHelperContent() {
      return {
        text: this.$gettext('Invite persons or groups to access this file or folder.')
      }
    },
    editDropDownToggleId() {
      return uuid.v4()
    }
  },
  methods: {
    ...mapActions(['showMessage']),
    ...mapActions('Files', ['changeShare']),

    removeShare() {
      this.$emit('onDelete', this.share)
    },

    showAccessDetails() {
      console.log('SHOW EVENT EMITTED')
      this.$refs.accessDetails.show()
    },

    shareRoleChanged({ role, permissions }) {
      const expirationDate = this.share.expires
      try {
        this.saveShareChanges({ role, permissions, expirationDate })
      } catch (e) {
        console.error(e)
        this.showMessage({
          title: this.$gettext('Failed to apply new permissions'),
          status: 'danger'
        })
      }
    },

    shareExpirationChanged({ expirationDate }) {
      const role = this.share.role
      const permissions = this.share.customPermissions
      try {
        this.saveShareChanges({ role, permissions, expirationDate })
      } catch (e) {
        console.error(e)
        this.showMessage({
          title: this.$gettext('Failed to apply expiration date'),
          status: 'danger'
        })
      }
    },

    saveShareChanges({ role, permissions, expirationDate }) {
      const bitmask = role.hasCustomPermissions
        ? SharePermissions.permissionsToBitmask(permissions)
        : SharePermissions.permissionsToBitmask(
            role.permissions(this.hasResharing || this.shareType === ShareTypes.space)
          )
      this.changeShare({
        client: this.$client,
        graphClient: this.graphClient,
        share: this.share,
        permissions: bitmask,
        expirationDate: expirationDate || '',
        role
      })
    }
  }
})
</script>

<style lang="scss" scoped>
.sharee-avatar {
  min-width: 48px;
}

.share-access-details-drop {
  dl {
    display: grid;
    grid-template-columns: max-content auto;
  }

  dt {
    grid-column-start: 1;
  }

  dd {
    grid-column-start: 2;
    margin-left: var(--oc-space-medium);
  }
}
</style>
