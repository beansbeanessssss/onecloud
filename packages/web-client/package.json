{
  "name": "@ownclouders/web-client",
  "version": "0.0.1",
  "description": "ownCloud web client",
  "license": "AGPL-3.0",
  "private": false,
  "author": "ownCloud GmbH <devops@owncloud.com>",
  "homepage": "https://github.com/owncloud/web/tree/master/packages/web-client",
  "repository": {
    "type": "git",
    "url": "https://github.com/owncloud/web",
    "directory": "packages/web-client"
  },
  "exports": {
    ".": {
      "default": "./src/index.ts"
    },
    "./graph": {
      "default": "./src/graph/index.ts"
    },
    "./graph/generated": {
      "default": "./src/graph/generated/index.ts"
    },
    "./ocs": {
      "default": "./src/ocs/index.ts"
    },
    "./sse": {
      "default": "./src/sse/index.ts"
    },
    "./webdav": {
      "default": "./src/webdav/index.ts"
    },
    "./l10n": {
      "default": "./l10n/translations.json"
    }
  },
  "scripts": {
    "generate-openapi": "rm -rf src/generated && docker run --rm -v \"${PWD}/src:/local\" openapitools/openapi-generator-cli generate -i https://raw.githubusercontent.com/owncloud/libre-graph-api/main/api/openapi-spec/v1.0.yaml  -g typescript-axios -o /local/generated"
  },
  "peerDependencies": {
    "@ownclouders/web-client": "workspace:*",
    "@casl/ability": "^6.3.3",
    "axios": "1.6.8",
    "fast-xml-parser": "4.3.3",
    "lodash-es": "^4.17.21",
    "luxon": "3.2.1",
    "@microsoft/fetch-event-source": "^2.0.1",
    "webdav": "5.3.1",
    "xml-js": "^1.6.11",
    "zod": "3.22.4"
  }
}
