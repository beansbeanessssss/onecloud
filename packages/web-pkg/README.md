# web-pkg

This is experimental.
