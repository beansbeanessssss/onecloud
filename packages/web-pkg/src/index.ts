export * from './components'
export * from './composables'
export * from './configuration'
export * from './errors'
export * from './helpers'
export * from './http'
export * from './observer'
export * from './services'
export * from './utils'
