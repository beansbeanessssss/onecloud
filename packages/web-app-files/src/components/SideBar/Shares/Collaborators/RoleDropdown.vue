<template>
  <span v-if="selectedRole" class="oc-flex oc-flex-middle">
    <span v-if="availableRoles.length === 1">
      <oc-icon v-if="showIcon" :name="selectedRole.icon" class="oc-mr-s" />
      <span v-text="inviteLabel" />
    </span>
    <div v-else v-oc-tooltip="dropButtonTooltip">
      <oc-button
        :id="roleButtonId"
        class="files-recipient-role-select-btn"
        appearance="raw"
        gap-size="none"
        :disabled="isLocked"
        :aria-label="
          mode === 'create' ? $gettext('Select permission') : $gettext('Edit permission')
        "
      >
        <oc-icon v-if="showIcon" :name="selectedRole.icon" class="oc-mr-s" />
        <span class="oc-text-truncate" v-text="inviteLabel" />
        <oc-icon name="arrow-down-s" />
      </oc-button>
    </div>
    <oc-drop
      v-if="availableRoles.length > 1"
      ref="rolesDrop"
      :toggle="'#' + roleButtonId"
      mode="click"
      padding-size="small"
      class="files-recipient-role-drop"
      offset="0"
      close-on-click
    >
      <oc-list
        class="files-recipient-role-drop-list"
        :aria-label="$gettext('Select role for the invitation')"
      >
        <li v-for="role in availableRoles" :key="role.id">
          <oc-button
            :id="`files-recipient-role-drop-btn-${getRoleName(role)}`"
            ref="roleSelect"
            justify-content="space-between"
            class="files-recipient-role-drop-btn oc-p-s"
            :class="{
              'oc-background-primary-gradient': isSelectedRole(role),
              selected: isSelectedRole(role)
            }"
            :appearance="isSelectedRole(role) ? 'raw-inverse' : 'raw'"
            :variation="isSelectedRole(role) ? 'primary' : 'passive'"
            @click="selectRole(role)"
          >
            <span class="oc-flex oc-flex-middle">
              <oc-icon :name="role.icon" class="oc-pl-s oc-pr-m" variation="inherit" />
              <role-item :role="role" />
            </span>
            <span class="oc-flex">
              <oc-icon v-if="isSelectedRole(role)" name="check" variation="inherit" />
            </span>
          </oc-button>
        </li>
      </oc-list>
    </oc-drop>
  </span>
</template>

<script lang="ts">
import get from 'lodash-es/get'
import { storeToRefs } from 'pinia'
import RoleItem from '../Shared/RoleItem.vue'
import * as uuid from 'uuid'
import {
  defineComponent,
  inject,
  PropType,
  ComponentPublicInstance,
  computed,
  ref,
  unref,
  Ref
} from 'vue'
import { useAbility, useUserStore } from '@ownclouders/web-pkg'
import { Resource } from '@ownclouders/web-client'
import { useGettext } from 'vue3-gettext'
import { useSharesStore } from '@ownclouders/web-pkg'
import { ShareRoleNG } from '@ownclouders/web-client/src/helpers'

export default defineComponent({
  name: 'RoleDropdown',
  components: { RoleItem },
  props: {
    existingRole: {
      type: Object as PropType<ShareRoleNG>,
      required: false,
      default: undefined
    },
    domSelector: {
      type: String,
      required: false,
      default: undefined
    },
    mode: {
      type: String,
      required: false,
      default: 'create'
    },
    showIcon: {
      type: Boolean,
      default: false
    },
    isLocked: {
      type: Boolean,
      default: false
    }
  },
  emits: ['optionChange'],
  setup(props, { emit }) {
    const ability = useAbility()
    const userStore = useUserStore()
    const { user } = storeToRefs(userStore)
    const sharesStore = useSharesStore()
    const { graphRoles } = storeToRefs(sharesStore)
    const { $gettext } = useGettext()

    const dropButtonTooltip = computed(() => {
      if (props.isLocked) {
        return $gettext('Resource is temporarily locked, unable to manage share')
      }

      return ''
    })

    const availableRoles = inject<Ref<ShareRoleNG[]>>('availableShareRoles')

    const initialSelectedRole = props.existingRole ? props.existingRole : unref(availableRoles)[0]
    const selectedRole = ref<ShareRoleNG>(initialSelectedRole)

    const isSelectedRole = (role: ShareRoleNG) => {
      return unref(selectedRole).id === role.id
    }

    const selectRole = (role: ShareRoleNG) => {
      selectedRole.value = role
      emit('optionChange', unref(selectedRole))
    }

    // FIXME: only needed for e2e and acceptance tests, map id to human readable element id
    const getRoleName = (role: ShareRoleNG) => {
      const map = {
        'b1e2218d-eef8-4d4c-b82d-0f1a1b48f3b5': 'viewer',
        'a8d5fe5e-96e3-418d-825b-534dbdf22b99': 'viewer',
        '2d00ce52-1fc2-4dbc-8b95-a73b73395f5a': 'editor',
        'fb6c3e19-e378-47e5-b277-9732f9de6e21': 'editor',
        '58c63c02-1d89-4572-916a-870abc5a1b7d': 'editor',
        '312c0871-5ef7-4b3a-85b6-0e4074c64049': 'manager',
        '1c996275-f1c9-4e71-abdf-a42f6495e960': 'uploader'
      }
      return map[role.id]
    }

    return {
      ability,
      user,
      dropButtonTooltip,
      resource: inject<Resource>('resource'),
      graphRoles,
      selectedRole,
      availableRoles,
      isSelectedRole,
      selectRole,
      getRoleName
    }
  },
  computed: {
    roleButtonId() {
      if (this.domSelector) {
        return `files-collaborators-role-button-${this.domSelector}-${uuid.v4()}`
      }
      return 'files-collaborators-role-button-new'
    },
    inviteLabel() {
      return this.$gettext(this.selectedRole?.label || '')
    }
  },

  beforeUnmount() {
    window.removeEventListener('keydown', this.cycleRoles)
  },

  mounted() {
    window.addEventListener('keydown', this.cycleRoles)
  },

  methods: {
    cycleRoles(event) {
      // events only need to be captured if the roleSelect element is visible
      if (!get(this.$refs.rolesDrop, 'tippy.state.isShown', false)) {
        return
      }

      const { code } = event
      const isArrowUp = code === 'ArrowUp'
      const isArrowDown = code === 'ArrowDown'

      // to cycle through the list of roles only up and down keyboard events are allowed
      // if this is not the case we can return early and stop the script execution from here
      if (!isArrowUp && !isArrowDown) {
        return
      }

      // if there is only 1 or no roleSelect we can early return
      // it does not make sense to cycle through it if value is less than 1
      const roleSelect = (this.$refs.roleSelect as ComponentPublicInstance[]) || []

      if (roleSelect.length <= 1) {
        return
      }

      // obtain active role select in following priority chain:
      // first try to get the focused select
      // then try to get the selected select
      // and if none of those applies we fall back to the first role select
      const activeRoleSelect =
        roleSelect.find((rs) => rs.$el === document.activeElement) ||
        roleSelect.find((rs) => rs.$el.classList.contains('selected')) ||
        roleSelect[0]
      const activeRoleSelectIndex = roleSelect.indexOf(activeRoleSelect)
      const activateRoleSelect = (idx) => roleSelect[idx].$el.focus()

      // if the event key is arrow up
      // and the next active role select index would be less than 0
      // then activate the last available role select
      if (isArrowUp && activeRoleSelectIndex - 1 < 0) {
        activateRoleSelect(roleSelect.length - 1)

        return
      }

      // if the event key is arrow down
      // and the next active role select index would be greater or even to the available amount of role selects
      // then activate the first available role select
      if (isArrowDown && activeRoleSelectIndex + 1 >= roleSelect.length) {
        activateRoleSelect(0)

        return
      }

      // the only missing part is to navigate up or down, this only happens if:
      // the next active role index is greater than 0
      // the next active role index is less than the amount of all available role selects
      activateRoleSelect(activeRoleSelectIndex + (isArrowUp ? -1 : 1))
    }
  }
})
</script>

<style lang="scss" scoped>
.files-recipient {
  &-role-drop {
    @media (max-width: $oc-breakpoint-medium-default) {
      width: 100%;
    }
    @media (min-width: $oc-breakpoint-medium-default) {
      width: 400px;
    }

    &-list {
      li {
        margin: var(--oc-space-xsmall) 0;

        &:first-child {
          margin-top: 0;
        }
        &:last-child {
          margin-bottom: 0;
        }
      }
    }

    &-btn {
      width: 100%;
      gap: var(--oc-space-medium);

      &:hover,
      &:focus {
        background-color: var(--oc-color-background-hover);
        text-decoration: none;
      }
    }
  }
  &-role-select-btn {
    max-width: 100%;
  }

  &-custom-permissions-drop-cancel-confirm-btns {
    background: var(--oc-color-background-hover);
    text-align: right;
  }
}

.files-collaborators-permission-checkbox::v-deep {
  .oc-checkbox {
    border: 2px solid var(--oc-color-border);
  }

  label {
    margin-left: var(--oc-space-small);
  }
}
</style>
