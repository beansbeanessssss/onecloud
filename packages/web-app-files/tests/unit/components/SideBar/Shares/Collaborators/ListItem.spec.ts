import ListItem from '../../../../../../src/components/SideBar/Shares/Collaborators/ListItem.vue'
import {
  CollaboratorShare,
  GraphSharePermission,
  ShareRoleNG,
  ShareTypes
} from '@ownclouders/web-client/src/helpers/share'
import {
  defaultPlugins,
  mount,
  defaultStubs,
  defaultComponentMocks,
  nextTicks,
  mockAxiosResolve
} from 'web-test-helpers'
import { useMessages, useSharesStore, useSpacesStore } from '@ownclouders/web-pkg'
import EditDropdown from '../../../../../../src/components/SideBar/Shares/Collaborators/EditDropdown.vue'
import RoleDropdown from '../../../../../../src/components/SideBar/Shares/Collaborators/RoleDropdown.vue'
import { mock } from 'vitest-mock-extended'
import { Resource, SpaceResource } from '@ownclouders/web-client'

vi.mock('@ownclouders/web-client/src/helpers/space/functions', () => ({
  buildSpace: vi.fn((space) => space)
}))

vi.mock('uuid', () => ({
  v4: () => {
    return '00000000-0000-0000-0000-000000000000'
  }
}))

const selectors = {
  userAvatarImage: 'avatar-image-stub.files-collaborators-collaborator-indicator',
  notUserAvatar: 'oc-avatar-item-stub.files-collaborators-collaborator-indicator',
  collaboratorAdditionalInfo: '.files-collaborators-collaborator-additional-info',
  collaboratorName: '.files-collaborators-collaborator-name',
  accessDetailsButton: '.files-collaborators-collaborator-access-details-button',
  collaboratorRole: '.files-collaborators-collaborator-role',
  collaboratorEdit: '.files-collaborators-collaborator-edit',
  shareInheritanceIndicators: '.oc-resource-indicators',
  expirationDateIcon: '[data-testid="recipient-info-expiration-date"]'
}

const getShareMock = ({
  sharedWith,
  shareType,
  expirationDateTime
}: Partial<CollaboratorShare> = {}): CollaboratorShare => ({
  id: '1',
  sharedWith: sharedWith || { id: '3', displayName: 'einstein' },
  sharedBy: { id: '2', displayName: 'marie' },
  permissions: [],
  shareType: shareType || ShareTypes.user.value,
  role: mock<ShareRoleNG>({ description: '', displayName: '' }),
  resourceId: '1',
  indirect: false,
  expirationDateTime: expirationDateTime || ''
})

describe('Collaborator ListItem component', () => {
  describe('displays the correct image/icon according to the shareType', () => {
    describe('user and spaceUser share type', () => {
      it.each([ShareTypes.user.value, ShareTypes.spaceUser.value])(
        'should display a users avatar',
        (shareType) => {
          const { wrapper } = createWrapper({ share: getShareMock({ shareType }) })
          expect(wrapper.find(selectors.userAvatarImage).exists()).toBeTruthy()
          expect(wrapper.find(selectors.notUserAvatar).exists()).toBeFalsy()
        }
      )
      it('sets user info on the avatar', () => {
        const share = getShareMock()
        const { wrapper } = createWrapper({ share })
        expect(wrapper.find(selectors.userAvatarImage).attributes('userid')).toEqual(
          share.sharedWith.id
        )
        expect(wrapper.find(selectors.userAvatarImage).attributes('user-name')).toEqual(
          share.sharedWith.displayName
        )
      })
    })
    describe('non-user share types', () => {
      it.each(
        ShareTypes.all.filter(
          (shareType) => ![ShareTypes.user, ShareTypes.spaceUser].includes(shareType)
        )
      )('should display an oc-avatar-item for any non-user share types', (shareType) => {
        const { wrapper } = createWrapper({ share: getShareMock({ shareType: shareType.value }) })
        expect(wrapper.find(selectors.userAvatarImage).exists()).toBeFalsy()
        expect(wrapper.find(selectors.notUserAvatar).exists()).toBeTruthy()
        expect(wrapper.find(selectors.notUserAvatar).attributes().name).toEqual(shareType.key)
      })
      it('should display an oc-avatar-item for space group shares', () => {
        const { wrapper } = createWrapper({
          share: getShareMock({
            shareType: ShareTypes.spaceGroup.value,
            sharedWith: { id: '1', displayName: 'someGroup' }
          })
        })
        expect(wrapper.find(selectors.userAvatarImage).exists()).toBeFalsy()
        expect(wrapper.find(selectors.notUserAvatar).exists()).toBeTruthy()
      })
    })
  })
  describe('share info', () => {
    it('shows the collaborator display name', () => {
      const share = getShareMock()
      const { wrapper } = createWrapper({ share })
      expect(wrapper.find(selectors.collaboratorName).text()).toEqual(share.sharedWith.displayName)
    })
    it('shows the share expiration date if given', () => {
      const { wrapper } = createWrapper({
        share: getShareMock({ expirationDateTime: '2000-01-01' })
      })
      expect(wrapper.find(selectors.expirationDateIcon).exists()).toBeTruthy()
    })
  })
  describe('modifiable property', () => {
    it('shows interactive elements when modifiable', () => {
      const { wrapper } = createWrapper({ modifiable: true })
      expect(wrapper.find(selectors.collaboratorRole).exists()).toBeTruthy()
    })
    it('hides interactive elements when not modifiable', () => {
      const { wrapper } = createWrapper({ modifiable: false })
      expect(wrapper.find(selectors.collaboratorRole).exists()).toBeFalsy()
    })
  })
  describe('share inheritance indicators', () => {
    it('show when sharedParentRoute is given', () => {
      const { wrapper } = createWrapper({
        sharedParentRoute: { params: { driveAliasAndItem: '/folder' } }
      })
      expect(wrapper.find(selectors.shareInheritanceIndicators).exists()).toBeTruthy()
      expect(wrapper.html()).toMatchSnapshot()
    })
    it('do not show when sharedParentRoute is not given', () => {
      const { wrapper } = createWrapper()
      expect(wrapper.find(selectors.shareInheritanceIndicators).exists()).toBeFalsy()
    })
  })
  describe('remove share', () => {
    it('emits the "removeShare" event', () => {
      const { wrapper } = createWrapper()
      wrapper.findComponent<typeof EditDropdown>('edit-dropdown-stub').vm.$emit('removeShare')
      expect(wrapper.emitted().onDelete).toBeTruthy()
    })
  })
  describe('change share role', () => {
    it('calls "changeShare" for regular resources', () => {
      const { wrapper } = createWrapper()
      wrapper.findComponent<typeof RoleDropdown>('role-dropdown-stub').vm.$emit('optionChange', {
        permissions: [GraphSharePermission.readBasic]
      })
      const sharesStore = useSharesStore()
      expect(sharesStore.updateShare).toHaveBeenCalled()
    })
    it('calls "upsertSpaceMember" for space resources', async () => {
      const resource = mock<SpaceResource>({ driveType: 'project' })
      const { wrapper } = createWrapper({
        share: getShareMock({ shareType: ShareTypes.spaceUser.value }),
        resource
      })
      wrapper.findComponent<typeof RoleDropdown>('role-dropdown-stub').vm.$emit('optionChange', {
        permissions: [GraphSharePermission.readBasic]
      })

      await nextTicks(4)

      const spacesStore = useSpacesStore()
      expect(spacesStore.upsertSpaceMember).toHaveBeenCalled()
    })
    it('shows a message on error', async () => {
      const resource = mock<SpaceResource>({ driveType: 'project' })
      vi.spyOn(console, 'error').mockImplementation(() => undefined)
      const { wrapper } = createWrapper()
      vi.spyOn(wrapper.vm, 'saveShareChanges').mockImplementation(() => {
        throw new Error()
      })
      wrapper.findComponent<typeof RoleDropdown>('role-dropdown-stub').vm.$emit('optionChange', {
        share: getShareMock({ shareType: ShareTypes.spaceUser.value }),
        resource
      })

      await nextTicks(4)

      const sharesStore = useSharesStore()
      expect(sharesStore.updateShare).not.toHaveBeenCalled()
      const messagesStore = useMessages()
      expect(messagesStore.showErrorMessage).toHaveBeenCalled()
    })
  })
  describe('change expiration date', () => {
    it('calls "changeShare" for regular resources', () => {
      const { wrapper } = createWrapper()
      wrapper
        .findComponent<typeof EditDropdown>('edit-dropdown-stub')
        .vm.$emit('expirationDateChanged', {
          shareExpirationChanged: new Date()
        })
      const sharesStore = useSharesStore()
      expect(sharesStore.updateShare).toHaveBeenCalled()
    })
    it('shows a message on error', () => {
      vi.spyOn(console, 'error').mockImplementation(() => undefined)
      const { wrapper } = createWrapper()
      vi.spyOn(wrapper.vm, 'saveShareChanges').mockImplementation(() => {
        throw new Error()
      })
      wrapper
        .findComponent<typeof EditDropdown>('edit-dropdown-stub')
        .vm.$emit('expirationDateChanged', {
          shareExpirationChanged: new Date()
        })
      const sharesStore = useSharesStore()
      expect(sharesStore.updateShare).not.toHaveBeenCalled()
      const messagesStore = useMessages()
      expect(messagesStore.showErrorMessage).toHaveBeenCalled()
    })
  })
})

function createWrapper({
  share = getShareMock(),
  modifiable = true,
  sharedParentRoute = null,
  resource = mock<Resource>()
} = {}) {
  const mocks = defaultComponentMocks()
  mocks.$clientService.graphAuthenticated.drives.getDrive.mockResolvedValue(
    mockAxiosResolve(undefined)
  )

  return {
    wrapper: mount(ListItem, {
      props: {
        share,
        modifiable,
        sharedParentRoute
      },
      global: {
        plugins: [...defaultPlugins()],
        mocks,
        provide: { ...mocks, resource },
        renderStubDefaultSlot: true,
        stubs: {
          ...defaultStubs,
          'oc-icon': true,
          'avatar-image': true,
          'router-link': true,
          'oc-info-drop': true,
          'oc-table-simple': true,
          'oc-tr': true,
          'oc-td': true,
          'oc-tag': true,
          'oc-pagination': true,
          'oc-avatar-item': true,
          'role-dropdown': true,
          'edit-dropdown': true,
          translate: false
        }
      }
    })
  }
}
