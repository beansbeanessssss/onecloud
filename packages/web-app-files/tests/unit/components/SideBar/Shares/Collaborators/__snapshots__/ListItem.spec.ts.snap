// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Collaborator ListItem component > share inheritance indicators > show when sharedParentRoute is given 1`] = `
"<div data-v-ccf14be2="" data-testid="collaborator-user-item-einstein" class="files-collaborators-collaborator oc-py-xs">
  <div data-v-ccf14be2="" class="oc-width-1-1 oc-flex oc-flex-middle files-collaborators-collaborator-details">
    <div data-v-ccf14be2="" class="oc-width-2-3 oc-flex oc-flex-middle">
      <div data-v-ccf14be2="">
        <avatar-image-stub data-v-ccf14be2="" userid="3" user-name="einstein" width="36" class="files-collaborators-collaborator-indicator"></avatar-image-stub>
      </div>
      <div data-v-ccf14be2="" class="oc-pl-s oc-text-truncate">
        <div data-v-ccf14be2="" class="oc-text-truncate"><span data-v-ccf14be2="" aria-hidden="true" class="files-collaborators-collaborator-name">einstein</span> <span data-v-ccf14be2="" class="oc-invisible-sr">Share receiver name: einstein</span></div>
        <div data-v-ccf14be2="">
          <div data-v-ccf14be2="" class="oc-flex oc-flex-nowrap oc-flex-middle">
            <role-dropdown-stub data-v-ccf14be2="" existingrole="undefined" domselector="1" mode="edit" showicon="false" islocked="false" class="files-collaborators-collaborator-role"></role-dropdown-stub>
          </div>
        </div>
      </div>
    </div>
    <div data-v-ccf14be2="" class="oc-flex oc-flex-middle oc-width-1-3 files-collaborators-collaborator-navigation">
      <div data-v-ccf14be2="" class="oc-resource-indicators oc-text-truncate">
        <router-link-stub data-v-ccf14be2="" class="parent-folder oc-text-truncate" to="[object Object]"><span data-v-ccf14be2="" class="text">via</span>
          <oc-icon-stub data-v-ccf14be2="" name="folder-2" size="small" fill-type="line" class="oc-px-xs"></oc-icon-stub> <span data-v-ccf14be2="" class="text oc-text-truncate">folder</span>
        </router-link-stub>
      </div>
      <!--v-if-->
      <edit-dropdown-stub data-v-ccf14be2="" sharecategory="user" caneditordelete="true" issharedenied="false" deniable="false" islocked="false" id="edit-drop-down-00000000-0000-0000-0000-000000000000" class="files-collaborators-collaborator-edit" data-testid="collaborator-edit"></edit-dropdown-stub>
      <oc-info-drop-stub data-v-ccf14be2="" class="share-access-details-drop" title="Access details" list="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" mode="manual" target="#edit-drop-down-00000000-0000-0000-0000-000000000000"></oc-info-drop-stub>
    </div>
  </div>
</div>"
`;
